File: tests/python/shared/geometry.py
# Copyright (c) Honua. All rights reserved.
# Licensed under the Elastic License 2.0. See LICENSE in the project root.

"""
Geometry generators for comprehensive spatial testing.

Supports all GeoJSON geometry types:
- Point, MultiPoint
- LineString, MultiLineString
- Polygon (with holes), MultiPolygon (with holes)
- GeometryCollection
- Null geometries
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from shapely.geometry import (
    LineString,
    MultiLineString,
    MultiPoint,
    Point,
    mapping,
    shape,
)


def _load_geometry_fixtures() -> dict[str, Any]:
    fixture_path = Path(__file__).resolve().parents[2] / "fixtures" / "geometry-fixtures.json"
    return json.loads(fixture_path.read_text(encoding="utf-8"))


_GEOMETRY_FIXTURES = _load_geometry_fixtures()
_FIXTURE_GEOMETRIES = _GEOMETRY_FIXTURES["geometries"]
_BASE_LON = _GEOMETRY_FIXTURES["base"]["lon"]
_BASE_LAT = _GEOMETRY_FIXTURES["base"]["lat"]

@dataclass
class TestGeometry:
    """A test geometry with metadata for validation."""

    __test__ = False

    name: str
    geojson: dict[str, Any]
    wkt: str
    geometry_type: str
    has_holes: bool = False
    is_multi: bool = False
    is_null: bool = False

    @property
    def shapely(self):
        """Get shapely geometry for validation."""
        if self.is_null:
            return None
        return shape(self.geojson)

    def to_esri_json(self) -> dict[str, Any] | None:
        """Convert to Esri JSON format for FeatureServer tests."""
        if self.is_null:
            return None
        return geojson_to_esri(self.geojson)


def geojson_to_esri(geojson: dict[str, Any]) -> dict[str, Any]:
    """Convert GeoJSON geometry to Esri JSON format."""
    geom_type = geojson.get("type", "")
    coords = geojson.get("coordinates", [])

    if geom_type == "Point":
        return {"x": coords[0], "y": coords[1], "spatialReference": {"wkid": 4326}}

    elif geom_type == "MultiPoint":
        return {
            "points": coords,
            "spatialReference": {"wkid": 4326},
        }

    elif geom_type == "LineString":
        return {
            "paths": [coords],
            "spatialReference": {"wkid": 4326},
        }

    elif geom_type == "MultiLineString":
        return {
            "paths": coords,
            "spatialReference": {"wkid": 4326},
        }

    elif geom_type == "Polygon":
        return {
            "rings": coords,
            "spatialReference": {"wkid": 4326},
        }

    elif geom_type == "MultiPolygon":
        # Flatten all rings from all polygons
        rings = []
        for polygon in coords:
            rings.extend(polygon)
        return {
            "rings": rings,
            "spatialReference": {"wkid": 4326},
        }

    elif geom_type == "GeometryCollection":
        # Esri doesn't support GeometryCollection directly
        # Return the first geometry or None
        geometries = geojson.get("geometries", [])
        if geometries:
            return geojson_to_esri(geometries[0])
        return None

    return {}


class GeometryGenerator:
    """
    Generates test geometries for comprehensive spatial coverage.

    Usage:
        gen = GeometryGenerator()
        for geom in gen.all_geometries():
            # Test with geom.geojson, geom.wkt, geom.shapely
            pass
    """

    # Base coordinates for test data (San Francisco area)
    BASE_LON = _BASE_LON
    BASE_LAT = _BASE_LAT

    def _from_fixture(self, key: str, name: str | None = None) -> TestGeometry:
        fixture = _FIXTURE_GEOMETRIES[key]
        geojson = fixture["geojson"]
        is_null = fixture.get("isNull", False) or geojson is None
        geometry_type = fixture.get("geometryType", "Unknown")
        has_holes = fixture.get("hasHoles", False)
        is_multi = fixture.get("isMulti", False)
        resolved_name = name or fixture.get("defaultName", key)

        wkt = "" if is_null else shape(geojson).wkt

        return TestGeometry(
            name=resolved_name,
            geojson=geojson,
            wkt=wkt,
            geometry_type=geometry_type,
            has_holes=has_holes,
            is_multi=is_multi,
            is_null=is_null,
        )

    def point(self, name: str = "test_point", lon: float = None, lat: float = None) -> TestGeometry:
        """Generate a Point geometry."""
        if lon is None and lat is None:
            return self._from_fixture("point", name)

        lon = lon or self.BASE_LON
        lat = lat or self.BASE_LAT
        geom = Point(lon, lat)
        return TestGeometry(
            name=name,
            geojson=mapping(geom),
            wkt=geom.wkt,
            geometry_type="Point",
        )

    def multipoint(self, name: str = "test_multipoint", count: int = 3) -> TestGeometry:
        """Generate a MultiPoint geometry."""
        if count == 3:
            return self._from_fixture("multipoint", name)

        points = [(self.BASE_LON + i * 0.01, self.BASE_LAT + i * 0.01) for i in range(count)]
        geom = MultiPoint(points)
        return TestGeometry(
            name=name,
            geojson=mapping(geom),
            wkt=geom.wkt,
            geometry_type="MultiPoint",
            is_multi=True,
        )

    def linestring(self, name: str = "test_linestring", points: int = 4) -> TestGeometry:
        """Generate a LineString geometry."""
        if points == 4:
            return self._from_fixture("linestring", name)

        coords = [(self.BASE_LON + i * 0.01, self.BASE_LAT + i * 0.005) for i in range(points)]
        geom = LineString(coords)
        return TestGeometry(
            name=name,
            geojson=mapping(geom),
            wkt=geom.wkt,
            geometry_type="LineString",
        )

    def multilinestring(self, name: str = "test_multilinestring", lines: int = 2) -> TestGeometry:
        """Generate a MultiLineString geometry."""
        if lines == 2:
            return self._from_fixture("multilinestring", name)

        all_lines = []
        for line_idx in range(lines):
            coords = [
                (self.BASE_LON + i * 0.01, self.BASE_LAT + line_idx * 0.02 + i * 0.005)
                for i in range(3)
            ]
            all_lines.append(coords)
        geom = MultiLineString(all_lines)
        return TestGeometry(
            name=name,
            geojson=mapping(geom),
            wkt=geom.wkt,
            geometry_type="MultiLineString",
            is_multi=True,
        )

    def polygon_simple(self, name: str = "test_polygon_simple") -> TestGeometry:
        """Generate a simple Polygon without holes."""
        return self._from_fixture("polygonSimple", name)

    def polygon_with_hole(self, name: str = "test_polygon_with_hole") -> TestGeometry:
        """Generate a Polygon with one hole."""
        return self._from_fixture("polygonWithHole", name)

    def polygon_with_multiple_holes(
        self, name: str = "test_polygon_multiple_holes"
    ) -> TestGeometry:
        """Generate a Polygon with multiple holes."""
        return self._from_fixture("polygonWithMultipleHoles", name)

    def multipolygon_simple(self, name: str = "test_multipolygon_simple") -> TestGeometry:
        """Generate a MultiPolygon without holes."""
        return self._from_fixture("multipolygonSimple", name)

    def multipolygon_with_holes(
        self, name: str = "test_multipolygon_with_holes"
    ) -> TestGeometry:
        """Generate a MultiPolygon with holes in some polygons."""
        return self._from_fixture("multipolygonWithHoles", name)

    def geometry_collection(self, name: str = "test_geometry_collection") -> TestGeometry:
        """Generate a GeometryCollection with mixed geometry types."""
        return self._from_fixture("geometryCollection", name)

    def null_geometry(self, name: str = "test_null_geometry") -> TestGeometry:
        """Generate a null geometry representation."""
        return self._from_fixture("nullGeometry", name)

    def all_geometries(self) -> list[TestGeometry]:
        """Generate all supported geometry types for comprehensive testing."""
        return [
            self.point(),
            self.multipoint(),
            self.linestring(),
            self.multilinestring(),
            self.polygon_simple(),
            self.polygon_with_hole(),
            self.polygon_with_multiple_holes(),
            self.multipolygon_simple(),
            self.multipolygon_with_holes(),
            self.geometry_collection(),
            self.null_geometry(),
        ]

    def points_grid(
        self,
        name_prefix: str = "grid",
        rows: int = 5,
        cols: int = 5,
        spacing: float = 0.01,
    ) -> list[TestGeometry]:
        """Generate a grid of point geometries."""
        geometries = []
        for r in range(rows):
            for c in range(cols):
                lon = self.BASE_LON + c * spacing
                lat = self.BASE_LAT + r * spacing
                name = f"{name_prefix}_{r}_{c}"
                geometries.append(self.point(name=name, lon=lon, lat=lat))
        return geometries

    def bbox(
        self, min_lon: float = None, min_lat: float = None, max_lon: float = None, max_lat: float = None
    ) -> tuple[float, float, float, float]:
        """Get a bounding box tuple (min_lon, min_lat, max_lon, max_lat)."""
        return (
            min_lon or self.BASE_LON - 0.1,
            min_lat or self.BASE_LAT - 0.1,
            max_lon or self.BASE_LON + 0.1,
            max_lat or self.BASE_LAT + 0.1,
        )


# Convenience list of all geometry type names
ALL_GEOMETRY_TYPES = _GEOMETRY_FIXTURES.get("geometryTypes", [])

File: tests/js/shared/geometry.ts
/**
 * Geometry generator for comprehensive spatial testing.
 *
 * Supports all GeoJSON geometry types with conversion to Esri JSON format.
 * Mirrors the Python GeometryGenerator for test consistency.
 */

import { readFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import type { Geometry, Point, MultiPoint, LineString, MultiLineString, Position } from 'geojson';

// =============================================================================
// Types
// =============================================================================

/** Esri JSON geometry types */
export interface EsriPoint {
  x: number;
  y: number;
  spatialReference: { wkid: number };
}

export interface EsriMultiPoint {
  points: Position[];
  spatialReference: { wkid: number };
}

export interface EsriPolyline {
  paths: Position[][];
  spatialReference: { wkid: number };
}

export interface EsriPolygon {
  rings: Position[][];
  spatialReference: { wkid: number };
}

export interface EsriEnvelope {
  xmin: number;
  ymin: number;
  xmax: number;
  ymax: number;
  spatialReference: { wkid: number };
}

export type EsriGeometry = EsriPoint | EsriMultiPoint | EsriPolyline | EsriPolygon | EsriEnvelope | null;

/** Test geometry with metadata */
export interface TestGeometry {
  name: string;
  geojson: Geometry | null;
  esriJson: EsriGeometry;
  geometryType: string;
  hasHoles: boolean;
  isMulti: boolean;
  isNull: boolean;
}

interface GeometryFixture {
  defaultName: string;
  geometryType: string;
  hasHoles: boolean;
  isMulti: boolean;
  isNull: boolean;
  geojson: Geometry | null;
}

interface GeometryFixtures {
  base: { lon: number; lat: number };
  geometries: Record<string, GeometryFixture>;
}

const fixturesPath = resolve(
  dirname(fileURLToPath(import.meta.url)),
  '../../fixtures/geometry-fixtures.json',
);
const geometryFixtures = JSON.parse(readFileSync(fixturesPath, 'utf8')) as GeometryFixtures;

// =============================================================================
// GeoJSON to Esri JSON Conversion
// =============================================================================

/**
 * Convert GeoJSON geometry to Esri JSON format.
 */
export function geojsonToEsri(geojson: Geometry | null): EsriGeometry {
  if (!geojson) return null;

  const wkid = 4326;

  switch (geojson.type) {
    case 'Point':
      return {
        x: geojson.coordinates[0],
        y: geojson.coordinates[1],
        spatialReference: { wkid },
      };

    case 'MultiPoint':
      return {
        points: geojson.coordinates,
        spatialReference: { wkid },
      };

    case 'LineString':
      return {
        paths: [geojson.coordinates],
        spatialReference: { wkid },
      };

    case 'MultiLineString':
      return {
        paths: geojson.coordinates,
        spatialReference: { wkid },
      };

    case 'Polygon':
      return {
        rings: geojson.coordinates,
        spatialReference: { wkid },
      };

    case 'MultiPolygon':
      // Flatten all rings from all polygons
      const rings: Position[][] = [];
      for (const polygon of geojson.coordinates) {
        rings.push(...polygon);
      }
      return {
        rings,
        spatialReference: { wkid },
      };

    case 'GeometryCollection':
      // Esri doesn't support GeometryCollection directly
      // Return the first geometry or null
      if (geojson.geometries.length > 0) {
        return geojsonToEsri(geojson.geometries[0] as Geometry);
      }
      return null;

    default:
      return null;
  }
}

// =============================================================================
// Geometry Generator
// =============================================================================

/**
 * Generates test geometries for comprehensive spatial coverage.
 *
 * Usage:
 * ```ts
 * const gen = new GeometryGenerator();
 * const point = gen.point();
 * const allGeoms = gen.allGeometries();
 * ```
 */
export class GeometryGenerator {
  // Base coordinates for test data (San Francisco area)
  private readonly baseLon = geometryFixtures.base.lon;
  private readonly baseLat = geometryFixtures.base.lat;

  private fromFixture(key: string, nameOverride?: string): TestGeometry {
    const fixture = geometryFixtures.geometries[key];
    if (!fixture) {
      throw new Error(`Unknown geometry fixture: ${key}`);
    }

    const name = nameOverride ?? fixture.defaultName;
    const geojson = fixture.geojson as Geometry | null;

    return {
      name,
      geojson,
      esriJson: geojsonToEsri(geojson),
      geometryType: fixture.geometryType,
      hasHoles: fixture.hasHoles,
      isMulti: fixture.isMulti,
      isNull: fixture.isNull,
    };
  }

  /**
   * Generate a Point geometry.
   */
  point(name = 'test_point', lon?: number, lat?: number): TestGeometry {
    if (lon === undefined && lat === undefined) {
      return this.fromFixture('point', name);
    }

    const x = lon ?? this.baseLon;
    const y = lat ?? this.baseLat;
    const geojson: Point = {
      type: 'Point',
      coordinates: [x, y],
    };
    return {
      name,
      geojson,
      esriJson: geojsonToEsri(geojson),
      geometryType: 'Point',
      hasHoles: false,
      isMulti: false,
      isNull: false,
    };
  }

  /**
   * Generate a MultiPoint geometry.
   */
  multipoint(name = 'test_multipoint', count = 3): TestGeometry {
    if (count === 3) {
      return this.fromFixture('multipoint', name);
    }

    const points: Position[] = [];
    for (let i = 0; i < count; i++) {
      points.push([this.baseLon + i * 0.01, this.baseLat + i * 0.01]);
    }
    const geojson: MultiPoint = {
      type: 'MultiPoint',
      coordinates: points,
    };
    return {
      name,
      geojson,
      esriJson: geojsonToEsri(geojson),
      geometryType: 'MultiPoint',
      hasHoles: false,
      isMulti: true,
      isNull: false,
    };
  }

  /**
   * Generate a LineString geometry.
   */
  linestring(name = 'test_linestring', pointCount = 4): TestGeometry {
    if (pointCount === 4) {
      return this.fromFixture('linestring', name);
    }

    const coords: Position[] = [];
    for (let i = 0; i < pointCount; i++) {
      coords.push([this.baseLon + i * 0.01, this.baseLat + i * 0.005]);
    }
    const geojson: LineString = {
      type: 'LineString',
      coordinates: coords,
    };
    return {
      name,
      geojson,
      esriJson: geojsonToEsri(geojson),
      geometryType: 'LineString',
      hasHoles: false,
      isMulti: false,
      isNull: false,
    };
  }

  /**
   * Generate a MultiLineString geometry.
   */
  multilinestring(name = 'test_multilinestring', lineCount = 2): TestGeometry {
    if (lineCount === 2) {
      return this.fromFixture('multilinestring', name);
    }

    const lines: Position[][] = [];
    for (let lineIdx = 0; lineIdx < lineCount; lineIdx++) {
      const coords: Position[] = [];
      for (let i = 0; i < 3; i++) {
        coords.push([
          this.baseLon + i * 0.01,
          this.baseLat + lineIdx * 0.02 + i * 0.005,
        ]);
      }
      lines.push(coords);
    }
    const geojson: MultiLineString = {
      type: 'MultiLineString',
      coordinates: lines,
    };
    return {
      name,
      geojson,
      esriJson: geojsonToEsri(geojson),
      geometryType: 'MultiLineString',
      hasHoles: false,
      isMulti: true,
      isNull: false,
    };
  }

  /**
   * Generate a simple Polygon without holes.
   */
  polygonSimple(name = 'test_polygon_simple'): TestGeometry {
    return this.fromFixture('polygonSimple', name);
  }

  /**
   * Generate a Polygon with one hole.
   */
  polygonWithHole(name = 'test_polygon_with_hole'): TestGeometry {
    return this.fromFixture('polygonWithHole', name);
  }

  /**
   * Generate a Polygon with multiple holes.
   */
  polygonWithMultipleHoles(name = 'test_polygon_multiple_holes'): TestGeometry {
    return this.fromFixture('polygonWithMultipleHoles', name);
  }

  /**
   * Generate a MultiPolygon without holes.
   */
  multipolygonSimple(name = 'test_multipolygon_simple'): TestGeometry {
    return this.fromFixture('multipolygonSimple', name);
  }

  /**
   * Generate a MultiPolygon with holes in some polygons.
   */
  multipolygonWithHoles(name = 'test_multipolygon_with_holes'): TestGeometry {
    return this.fromFixture('multipolygonWithHoles', name);
  }

  /**
   * Generate a GeometryCollection with mixed geometry types.
   */
  geometryCollection(name = 'test_geometry_collection'): TestGeometry {
    return this.fromFixture('geometryCollection', name);
  }

  /**
   * Generate a null geometry representation.
   */
  nullGeometry(name = 'test_null_geometry'): TestGeometry {
    return this.fromFixture('nullGeometry', name);
  }

  /**
   * Generate all supported geometry types for comprehensive testing.
   */
  allGeometries(): TestGeometry[] {
    return [
      this.point(),
      this.multipoint(),
      this.linestring(),
      this.multilinestring(),
      this.polygonSimple(),
      this.polygonWithHole(),
      this.polygonWithMultipleHoles(),
      this.multipolygonSimple(),
      this.multipolygonWithHoles(),
      this.geometryCollection(),
      this.nullGeometry(),
    ];
  }

  /**
   * Generate a grid of point geometries for pagination testing.
   */
  pointsGrid(
    namePrefix = 'grid',
    rows = 5,
    cols = 5,
    spacing = 0.01,
  ): TestGeometry[] {
    const geometries: TestGeometry[] = [];
    for (let r = 0; r < rows; r++) {
      for (let c = 0; c < cols; c++) {
        const lon = this.baseLon + c * spacing;
        const lat = this.baseLat + r * spacing;
        const name = `${namePrefix}_${r}_${c}`;
        geometries.push(this.point(name, lon, lat));
      }
    }
    return geometries;
  }

  /**
   * Get a bounding box tuple (xmin, ymin, xmax, ymax).
   */
  bbox(
    minLon?: number,
    minLat?: number,
    maxLon?: number,
    maxLat?: number,
  ): [number, number, number, number] {
    return [
      minLon ?? this.baseLon - 0.1,
      minLat ?? this.baseLat - 0.1,
      maxLon ?? this.baseLon + 0.1,
      maxLat ?? this.baseLat + 0.1,
    ];
  }

  /**
   * Get an Esri envelope from a bounding box.
   */
  envelope(
    minLon?: number,
    minLat?: number,
    maxLon?: number,
    maxLat?: number,
  ): EsriEnvelope {
    const [xmin, ymin, xmax, ymax] = this.bbox(minLon, minLat, maxLon, maxLat);
    return {
      xmin,
      ymin,
      xmax,
      ymax,
      spatialReference: { wkid: 4326 },
    };
  }

  /**
   * Get a geometry by method name for matrix testing.
   */
  getByMethod(method: string): TestGeometry {
    const methodMap: Record<string, () => TestGeometry> = {
      point: () => this.point(),
      multipoint: () => this.multipoint(),
      linestring: () => this.linestring(),
      multilinestring: () => this.multilinestring(),
      polygonSimple: () => this.polygonSimple(),
      polygonWithHole: () => this.polygonWithHole(),
      polygonWithMultipleHoles: () => this.polygonWithMultipleHoles(),
      multipolygonSimple: () => this.multipolygonSimple(),
      multipolygonWithHoles: () => this.multipolygonWithHoles(),
      geometryCollection: () => this.geometryCollection(),
      nullGeometry: () => this.nullGeometry(),
    };

    const fn = methodMap[method];
    if (!fn) {
      throw new Error(`Unknown geometry method: ${method}`);
    }
    return fn();
  }
}

// Default export for convenience
export const geometryGenerator = new GeometryGenerator();
